`DataFrame.astype('category')` currently not supported

Weird, using v1.5.3.230227, which has that, I got this message

`
error: Argument 1 to "astype" of "DataFrame" has incompatible type "Literal['category']"; expected "Union[Union[Union[Type[bool], Type[bool_], BooleanDtype, Literal['bool']], Union[Type[int], Int8Dtype, Int16Dtype, Int32Dtype, Int64Dtype, Type[signedinteger[_8Bit]], Type[signedinteger[_16Bit]], Type[signedinteger[_32Bit]], Type[signedinteger[_64Bit]], Type[unsignedinteger[_8Bit]], Type[unsignedinteger[_16Bit]], Type[unsignedinteger[_32Bit]], Type[unsignedinteger[_64Bit]], Type[signedinteger[Any]], Type[unsignedinteger[Any]], Type[signedinteger[Any]], Type[unsignedinteger[Any]], Literal['int', 'int32']], Union[Type[str], StringDtype, Literal['str']], Type[bytes], Union[Float32Dtype, Float64Dtype, Type[floating[_16Bit]], Type[floating[_32Bit]], Type[floating[_64Bit]], Type[float], Literal['float']], Union[Type[complexfloating[_32Bit, _32Bit]], Type[complexfloating[_64Bit, _64Bit]], Type[complex], Literal['complex']], CategoricalDtype, ExtensionDtype, Literal['timedelta64[ns]', 'datetime64[ns]']], Mapping[Any, Union[ExtensionDtype, Union[str, dtype[generic], Type[str], Type[complex], Type[bool], Type[object]]]], Series[Any]]"  [arg-type]
`

I guess because `CategoryDtypeArg` is missing from `AstypeArg`.

https://github.com/pandas-dev/pandas-stubs/blob/d2faa7fb10fbed0303e33260f9aa7374974274ad/pandas-stubs/_typing.pyi#L153-L164

_Originally posted by @zmoon in https://github.com/pandas-dev/pandas-stubs/issues/556#issuecomment-1450624426_

--- a/pandas-stubs/_typing.pyi
+++ b/pandas-stubs/_typing.pyi
@@ -160,6 +160,7 @@ AstypeArg: TypeAlias = (
     | TimedeltaDtypeArg
     | TimestampDtypeArg
     | CategoricalDtype
+    | CategoryDtypeArg
     | ExtensionDtype
 )
 # DtypeArg specifies all allowable dtypes in a functions its dtype argument

--- a/pandas-stubs/_typing.pyi
+++ b/pandas-stubs/_typing.pyi
@@ -148,7 +148,7 @@ ComplexDtypeArg: TypeAlias = (
 )
 TimedeltaDtypeArg: TypeAlias = Literal["timedelta64[ns]"]
 TimestampDtypeArg: TypeAlias = Literal["datetime64[ns]"]
-CategoryDtypeArg: TypeAlias = Literal["category"]
+CategoryDtypeArg: TypeAlias = CategoricalDtype | Literal["category"]
 
 AstypeArg: TypeAlias = (
     BooleanDtypeArg
@@ -159,7 +159,6 @@ AstypeArg: TypeAlias = (
     | ComplexDtypeArg
     | TimedeltaDtypeArg
     | TimestampDtypeArg
-    | CategoricalDtype
     | CategoryDtypeArg
     | ExtensionDtype
 )

--- a/tests/test_frame.py
+++ b/tests/test_frame.py
@@ -2442,3 +2442,5 @@ def test_astype() -> None:
         assert_type(ab.astype({"col1": "int32", "col2": str}), "pd.DataFrame"),
         pd.DataFrame,
     )
+    check(assert_type(s.astype(pd.CategoricalDtype()), "pd.DataFrame"), pd.DataFrame)
+    check(assert_type(s.astype("category"), "pd.DataFrame"), pd.DataFrame)  # GH 559

--- a/tests/test_series.py
+++ b/tests/test_series.py
@@ -1596,10 +1596,10 @@ def test_updated_astype() -> None:
     check(
         assert_type(s.astype(pd.CategoricalDtype()), "pd.Series[Any]"),
         pd.Series,
-        pd.CategoricalDtype,
+        np.integer,
     )
     check(
         assert_type(s.astype("category"), "pd.Series[Any]"),
         pd.Series,
-        pd.CategoricalDtype,
+        np.integer,
     )

--- a/tests/test_series.py
+++ b/tests/test_series.py
@@ -1591,3 +1591,15 @@ def test_updated_astype() -> None:
         pd.Series,
         Decimal,
     )
+
+    # Categorical
+    check(
+        assert_type(s.astype(pd.CategoricalDtype()), "pd.Series[Any]"),
+        pd.Series,
+        pd.CategoricalDtype,
+    )
+    check(
+        assert_type(s.astype("category"), "pd.Series[Any]"),
+        pd.Series,
+        pd.CategoricalDtype,
+    )

